### ISSUE TITLE
Reconsidering the use of vue-async-computed

### ISSUE THREAD

**Issue by jbeezley** [girder/girder_web_components#39]
The [vue-async-computed plugin](https://github.com/foxbenjaminfox/vue-async-computed) was added by #35 in response to [my comment](https://github.com/girder/girder_web_components/pull/17#discussion_r222015550) in a different PR.  A [discussion](https://github.com/girder/girder_web_components/pull/36#discussion_r224081467) started in a more recent PR about some of the pitfalls of this approach.  I created this issue for a more focused discussion.  Here is a related issue with a response from the author that is worth reading: https://github.com/foxbenjaminfox/vue-async-computed/issues/30.

One of the metrics I use when evaluating how to write a Vue component is to minimize the use of functions that need to be called manually to update state; doing so largely nullifies the nice things that Vue provides for you.  In an ideal world, all state is managed by Vue and all of this happens automatically, but in real application there is always state that exists elsewhere.  For server side state, I think vue-async-computed provides a nice pattern for managing the duplication.  The alternative is adding `watch`'ed properties that spawn rest requests and set other derived properties when they return. 
 This essentially what vue-async-computed is doing for you anyway.

In regards to the use of a `refresh` counter in #36, I did a similar thing in [GirderJobList](https://github.com/Kitware/resonantgeo/blob/master/src/components/girder/GirderJobList.vue).  The primary difference is that in my use case, the counter is an internal detail which makes more palatable IMO.  I haven't spent too much time looking at the implementation in #36, but I suspect it could be handled instead by having a source of truth in Vue representing the server state that we *actually* want to watch.

I'm certainly willing to reconsider using it we decide it doesn't work well for us, but I think we should decide on a pattern for sync'ing server state to replace it.

**Comment by zachmullen** [comment]
+1, we definitely need a pattern that we can use everywhere. I think `vue-async-computed` is really nice for cases where we don't have to manually refresh. In cases where we do need to do so, if we do go the `refresh` route, it should be a method rather than a prop IMO:

```pug
girder-file-browser(ref="browser")
girder-upload(@done="$refs.browser.refresh()")
```

It's definitely a drag that `asyncComputed` properties can't be `set`, so I'm certainly open to other patterns, but we should also consider the risk in using `vue-async-computed`; if we can keep its use as an internal detail of this library, it strongly mitigates the risk of moving forward with it.

**Comment by jeffbaumes** [comment]
@jbeezley thanks for the pointer to the author's explanation. I agree with the comment mostly, but it implies that in the case of GET, there is no downside of `vue-async-computed`, but I'd argue that even then if the server state changed you can run into issues, as we do with the data browser. But I see the point that `vue-async-computed` could very well have many more upsides compared to this one possible downside, and I'd highly regard actual experience with that tradeoff when writing real code like ResonantGeo and the file browser, as opposed to someone (me) coming in with more philosophical concerns.

There are going to be tradeoffs upon tradeoffs between practicality and purity/abstraction. For example, a framework like Vuex seeks to enforce layers around mutation and REST calls, so that async server stuff is all done in one layer that is not mixed in with component code, but we are not thinking of going there. For ultra-purity, there would be a list of contents as a file browser prop or Vuex store, and the upload widget would return the delta of new objects on completion, which would be merged into the prop/store and cause updates on the UI - no async calls needed.

If we are not seeking to live in such an idealized world (which I recommend we don't try to do), then we should indeed make tradeoffs, and a refresh pattern to force updates, if standardized and communicated, could be one of those tradeoffs.

**Comment by jeffbaumes** [comment]
@zachmullen I think it needs to be a data property because that is the only way for the "set" to be called, though the implementation of refresh() could increment the prop to keep it more hidden, and make the data property named something that is private by convention, a la https://medium.com/vuejs-tips/tip-14-private-data-in-vue-js-3188b9f76f30:

> If you need to create private properties, you have to choose a different name, like editableValue for example. Other suggestion is to use _ and $ as suffix, like value_ or value$.

**Comment by subdavis** [comment]
+1 to internalizing the counter with a private name and incrementing it with a refresh function.  

* Is this a suitable solution for force refresh in the context of this library? 
* Should this decision apply to future cases where force refresh is needed?

**Comment by jbeezley** [comment]
Yes, I think the Vue-sanctioned way to deal with server communication is to use Vuex.  I initially tried to come up with a way to use Vuex modules, but didn't manage to make it work.  In the end, I came at the decision to use vue-async-computed because my cost function emphasizes minimizing the number of times `watch` is used manually.  To me, that is code smell indicating duplicate state.

**Comment by zachmullen** [comment]
Should we be using `shouldUpdate` instead of `watch`? That makes it easier to use, e.g. a boolean value for whether we should refresh, rather than a counter.

https://github.com/foxbenjaminfox/vue-async-computed#conditional-recalculation

**Comment by jbeezley** [comment]
If I understand correctly, `shouldUpdate` only allows you to prevent a refresh from occurring rather than forcing one.

**Comment by zachmullen** [comment]
Where did we end up on this, what are people's preferences? IMO we should make a decision on this and close this issue out.

The fundamental issue is that the authoritative state of the application necessarily:

* Is shared between multiple components
* Is external to the entire application (i.e. it's on the server)

so **we require a manual sync function somewhere or another.**

From a purity perspective, Vuex seems to be the best solution since it just separates out the state management completely, but I'm not sure how that works in a component library context. If anyone is arguing for Vuex as the solution, I think they should create an example in code of how it would look to both developers and downstreams of this library.

I believe that a lack of purity/consistency in addressing this problem was a major pain point in the Backbone client: do I pass in an ID and fetch inside the component, or fetch before I render the component and just pass in the resulting data? That's really an important question for us to consider. From a component library perspective, which makes more sense? That is, should we even attempt to address the sync problem in our components, or should it be left completely to downstreams?

**Comment by zachmullen** [comment]
We could imagine a library where all we do is present data and emit events, leaving synchronization with the server completely up to downstreams. That is, we would never use `girderRest` internal to our library components. This gives us great purity and simplicity, but provides decidedly less value as that burden of synchronization is now pushed downstream.

**Comment by jbeezley** [comment]
I agree, this is pretty critical as it affects the API surface.  I think the fact that we are already feeling pain here is troubling, and perhaps that is enough evidence to go in another direction.

I'm not totally convinced vue-async-computed is the core of the problem.  I took another look at the vue forums referencing reusable components w.r.t. server side state.  Just about every post references using vuex to manage server-side state.  I found the following [comment](https://forum.vuejs.org/t/generic-reusable-data-bound-custom-components-with-vuex/1828/2) from a vue maintainer:
```
There’s always two kinds of components:

A. “smart” components:

* They orchestrate other components,
* do very little in terms of UI handling
* handle the state (here, vuex is handled)
* not really re-usable, since they are specific to your app’s individual requirements / business logic.

B. “dumb” components

* only receive props.
* don’t know about the global state (vuex)
* take care of all the UI events (clicks, inputs …)
* highly re-usable because they represent re-occuring UI interactions (i.e. entering an address) and only depend on props
```
I think we've discussed these distinctions before, but it's interesting to have an authoritative reference.

Are we just rediscovering the fact that components that make server requests aren't reusable?  I know @zachmullen made this distinction in girder-vue, but that there was push back for the fact that it effectively doubled the number of components that need to exist.  There is also the problem that a lot (maybe even most?) of the complicated logic is in the client/server communication that isn't part of reusable code.  The presentation after all is mostly just vuetify at this point.

To answer the question in this PR, I suspect that removing vue-async-computed will bring a whole other class of problems.  I also don't feel like I have enough experience maintaining a sufficiently large application where these problems begin to surface, so I'm happy to go with either approach.

**Comment by zachmullen** [comment]
Is there a nice way we can make a *modular* Vuex store as part of this library? I realize this is Google-able, but I'm hoping someone on the thread may have actually tried it.

**Comment by jbeezley** [comment]
There are vuex modules, but it is unclear on how to reuse them.  There are several related questions on the vue forums that are all unanswered, .i.e.:

https://forum.vuejs.org/t/vuex-and-reusable-module-component-pairs/13539
https://forum.vuejs.org/t/each-component-instance-own-vuex-module/3306
https://forum.vuejs.org/t/help-understanding-reusable-components-in-vuex/1223

It seems like the kind of thing that people just don't do.

**Comment by jeffbaumes** [comment]
I agree with @jbeezley it's probably better to use async-computed than fending for ourselves without it. We just need to learn how exactly it works and when to use it (and not use it). I'd be fine with allowing its use.

The larger open question is the overall structure of this repo and the types of components it contains. @matthewma7 is already using some components from here I believe, and could maybe comment on the ease of connecting this repo's business logic with his app's business logic. Is it easy, or would he rather have "dumb" components instead and do the business logic himself?

**Comment by zachmullen** [comment]
I will say from experience on girder-vue that doing the business logic for the upload component specifically is a fair bit of the work; having it be internal to the component will save downstreams a lot of hassle.

**Comment by matthewma7** [comment]
About vuex module, I implemented a Prompt component that has a counterpart vuex module. It doesn't feel so natural to me at this moment, but definitely works (reusable in two applications)

**Comment by zachmullen** [comment]
Thanks for the insight @matthewma7, in particular your story of using this library.

For the Upload component, which was the first to be developed, we required the destination Object to be passed as a prop, rather than accepting an ID and fetching it ourselves. So, the Upload component makes XHRs for uploading, but does not do any reads from the server itself, i.e. it has no asyncComputed properties / GET requests. The Authentication component, on the other hand, does a XHR GET at init time to fetch the list of OAuth providers from the server. Are these two cases fundamentally different? (FileBrowser accepts an ID and fetches the data itself, so it's more like the Authentication component). I'm trying to figure out whether one approach can be universally prescribed, or whether each component should decide what to accept on a case-by-case basis driven by how we anticipate it will be used downstream.

For the case of building a generic Girder web client, I know that a very common thing will be to take a route containing an ID string and use it to display some components with the corresponding data. In that case, I think I'd prefer to pass the ID to the component rather than having to always fetch it myself first, just to keep my router layer extremely thin. I think it's a tradeoff between power and ease of use.

**Comment by matthewma7** [comment]
The Upload component is asking for a dest prop, I later found out that, strictly speaking, an object with name, _id, and _modelType, which commonly means an item or a folder model, but I can modify the name to custom name if I need to, which I found flexible.
So in Girder context, if the id (and type?) vs model is common enough, maybe we could make them universally support both without too much development effort? Yes, I do see the simplicity of passing an ID and the network efficiency and customizability of passing a model. It really is hard to say one way is better than the other.
